### PR TITLE
Added ValueTuple so we can start taking advantage of C# 7 tuples.

### PR DESCRIPTION
**Pull request by 333fred**
@agocke added ValueTuple as a core reference to Roslyn, so it should be in VS now. Tagging @dotnet/project-system for review.

### CODE REVIEW

**Comment by davkean** [comment]
This dependencies indicates our Roslyn dependency, ValueTuple isn't a Roslyn dependency.

**Comment by 333fred** [comment]
@davkean it seemed the best fit of the current dependency projects to use. I can make a new project for it if we want to.

**Comment by srivatsn** [comment]
A new project seems cleaner because then if a project wants to use tuples but doesn't need Roslyn APIs it can simply add a reference to the ValueTuple project.

**Comment by nguerrera** [comment]
As a middle ground, should we have a corefx dependency project?

File: src/Dependencies/Roslyn/project.json [hunk 1]
```diff
@@ -1,7 +1,8 @@
 {
   "supports": {},
   "dependencies": {
-    "Microsoft.VisualStudio.LanguageServices": "2.0.0-beta5-60909-09"
+    "Microsoft.VisualStudio.LanguageServices": "2.0.0-beta5-60909-09",
+    "System.ValueTuple": "4.0.1-beta-24425-02"
```
Comment (nguerrera) [inline src/Dependencies/Roslyn/project.json]: Why is this on a pre-release? There are stable versions available.

Comment (nguerrera) [inline src/Dependencies/Roslyn/project.json]: I think we should make sure the stable 4.3.0 is being used by VS and use that and not add another thing that needs to change between now and RTM for that.

**Comment by nguerrera** [comment]
Dependency projects are being removed from roslyn: https://github.com/dotnet/roslyn/pull/15465

Should we be doing the same?

**Comment by davkean** [comment]
RE Dependency projects - until we have something better (ie msbuild-based packagerefs that are unified in a targets), I still prefer them over Jared's approach which is to duplicate the refs at all the roots and than have an after the fact enforcement (read: PowerShell script) that makes sure they are the same version.

I can live what you've got here - given we're going to replace this with packagerefs anyway soon.

**Comment by davkean** [comment]
@333fred Can you figure out if this is supposed to be a RTM dependency and if not just merge it in?

**Comment by srivatsn** [comment]
There's a 4.3.0 of the ValueTuple package which is also the version that the IDE's codefix adds. Let's use that.

**Comment by 333fred** [comment]
@srivatsn, updated to 4.3.0. @jaredpar or @agocke, can either of you confirm that this is the final version?

**Comment by jaredpar** [comment]
CC @jcouv 

It's the final version until we make another change :)

**Comment by srivatsn** [comment]
We just need some stable dependency - so 4.3.0 is good. We can merge this change.

**Comment by abpiskunov** [comment]
This construct does not work: new (IDependencyNode a, IProjectDependenciesSubTreeProvider b) { }; how do i create a new Tuple with new syntax? Also be careful with var (type1 x , type2 y) = ...  it would throw if any of the values is null.

**Comment by srivatsn** [comment]
you can simply say `var (x, y) = (new DependencyNode(), new SubTreeProvider())`

**Comment by abpiskunov** [comment]
what if i don't want to instantiate them yet, with possible null values? I need to declare types not instances

**Comment by 333fred** [comment]
`(IDependencyNode node, IProjectDependenciesSubTreeProvider provider) = (null, null);`

**Comment by abpiskunov** [comment]
it does not work -red squiggles

**Comment by srivatsn** [comment]
What do the red squiggles say? - I just tried this on my machine and it works fine.

**Comment by 333fred** [comment]
@abpiskunov, try running a NuGet restore. I was seeing System.ValueTuple`2 not imported until I reran restore.

**Comment by mavasani** [comment]
I am trying to use value tuple in one of the property page providers and am able to build fine, but doing an F5 with ProjectSystemDogfoodSetup as the startup project generates the following FileNotFoundException when attempting to open property pages:

image

Do I need to move to a newer VS drop or is there a workaround?

**Comment by srivatsn** [comment]
The version number looks wrong - it should be 4.3.0.0. Maybe you have some stale version?

**Comment by davkean** [comment]
The version of ValueTuple in my install is 4.0.1.0 under PrivateAssemblies.

**Comment by mavasani** [comment]
Yes, the NuGet package version is 4.3.0 but the assembly is 4.0.1.0. I verified the assembly is missing in my PrivateAssemblies, so time to move to a newer D15PreRel drop for me :) Thanks!

**Comment by mavasani** [comment]
Btw, also worked by manually copying the assembly to PrivateAssemblies.

**Comment by srivatsn** [comment]
@jcouv has the 4.3.0 version been inserted to the latest PreRel builds?

**Comment by davkean** [comment]
Also filed this: https://github.com/dotnet/roslyn/issues/16036 as it looks like it's been shipped many times.

**Comment by abpiskunov** [comment]
@srivatsn after i re-ran nuget restore it started to work , will update my change in the next coming bug fix.

**Comment by jcouv** [comment]
@srivatsn Sorry, I don't understand the question.
My understanding is that a copy of System.ValueTuple.dll is copied in the VS install for the interactive window (originally added as 4.3-preview1 in https://github.com/dotnet/roslyn/pull/14376). 
And then Roslyn also has a reference to the nuget package (updated to 4.3 in https://github.com/dotnet/roslyn/pull/14879).

**Pull request merged**